crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -343,26 +354,41 @@ impl HighlightedRangeStack {
             if let Some(idx) =
                 prev.iter().position(|parent| parent.range.contains_range(child.range))
             {
+                if let Some(tag) = overwrite_parent {
+                    prev[idx].highlight = tag;
+                }
+
                 let cloned = Self::intersect(&mut prev[idx], &child);
-                let insert_idx = if delete || prev[idx].range.is_empty() {
+                let insert_idx = if prev[idx].range.is_empty() {
                     prev.remove(idx);
                     idx
                 } else {
                     idx + 1
                 };
                 prev.insert(insert_idx, child);
-                if !delete && !cloned.range.is_empty() {
+                if !cloned.range.is_empty() {
                     prev.insert(insert_idx + 1, cloned);
                 }
-            } else if let Some(_idx) =
-                prev.iter().position(|parent| parent.range.contains(child.range.start()))
-            {
-                unreachable!("child range should be completely contained in parent range");
             } else {
-                let idx = prev
-                    .binary_search_by_key(&child.range.start(), |range| range.range.start())
-                    .unwrap_or_else(|x| x);
-                prev.insert(idx, child);
+                let maybe_idx =
+                    prev.iter().position(|parent| parent.range.contains(child.range.start()));
+                if let (Some(_), Some(idx)) = (overwrite_parent, maybe_idx) {
```

Comment:
Have you observed this situation? I am pretty sure this can't happen in the current form. If a child-range only partially overlaps a parent range this has to be a bug

Reply:
I ran into this after ensuring `cloned` was added into `prev`. This was specifically the case with the mult-iline comment. I'm not sure if it's that multi-line comments also include the newline while single-line comments don't. Before this change, the child would be completely outside of the parent range, so it was possible to just add it to `prev`. Now because we're tracking the `cloned` area from the original doc-comment split, there's a mismatch at the very end (`cloned` from the original split is shorter than `child`).

Reply:
Alright, can you transform the if into a match? That would probably make it a bit more readable

Reply:
Done